## Data Understand and Data Preparation

 This project focus on data understanding and data preparation Data Understanding is the process of collecting, exploring and conducting quality check on data to ensure they are relevant to the particular problem at hand.


Question: "Can we automate the process of determing the cusine of a given recipe?"


Step 1: Downloadin the library dependencies:

In [2]:
import pandas as pd # import library to read data info dataframe
pd.set_option('display.max_columns', None) # display all columns in dataframe
import numpy as np # import numpy library
import re # import library for regular expressions
import requests



Step 2: Download the data need from IBM

In [3]:
recipes = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0103EN-SkillsNetwork/labs/Module%202/recipes.csv")

In [4]:
import requests


url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0103EN-SkillsNetwork/labs/Module%202/recipes.csv'
request = requests.get(url)
with open('recipes.csv', 'wb') as file:
    file.write(request.content)

recipes = pd.read_csv("recipes.csv")

In [6]:
recipes.head()

country almond angelica anise anise_seed apple apple_brandy apricot  \
0  Vietnamese     No       No    No         No    No           No      No   
1  Vietnamese     No       No    No         No    No           No      No   
2  Vietnamese     No       No    No         No    No           No      No   
3  Vietnamese     No       No    No         No    No           No      No   
4  Vietnamese     No       No    No         No    No           No      No   

  armagnac artemisia artichoke asparagus avocado bacon baked_potato balm  \
0       No        No        No        No      No    No           No   No   
1       No        No        No        No      No    No           No   No   
2       No        No        No        No      No    No           No   No   
3       No        No        No        No      No    No           No   No   
4       No        No        No        No      No    No           No   No   

  banana barley bartlett_pear basil bay bean beech beef beef_broth beef_liver  \
0     No     No            No   Yes  No   No    No   No         No         No   
1     No     No            No    No  No   No    No   No         No         No   
2     No     No            No    No  No   No    No   No         No         No   
3     No     No            No   Yes  No  Yes    No   No        Yes         No   
4     No     No            No    No  No   No    No   No         No         No   

  beer beet bell_pepper bergamot berry bitter_orange black_bean black_currant  \
0   No   No          No       No    No            No         No            No   
1   No   No          No       No    No            No         No            No   
2   No   No          No       No    No            No         No            No   
3   No   No          No       No    No            No         No            No   
4   No   No          No       No    No            No         No            No   

  black_mustard_seed_oil black_pepper black_raspberry black_sesame_seed  \
0                     No           No              No                No   
1                     No          Yes              No                No   
2                     No           No              No                No   
3                     No           No              No                No   
4                     No           No              No                No   

  black_tea blackberry blackberry_brandy blue_cheese blueberry bone_oil  \
0        No         No                No          No        No       No   
1        No         No                No          No        No       No   
2        No         No                No          No        No       No   
3        No         No                No          No        No       No   
4        No         No                No          No        No       No   

  bourbon_whiskey brandy brassica bread broccoli brown_rice brussels_sprout  \
0              No     No       No    No       No         No              No   
1              No     No       No    No       No         No              No   
2              No     No       No    No       No         No              No   
3              No     No       No    No       No         No              No   
4              No     No       No    No       No         No              No   

  buckwheat butter buttermilk cabbage cabernet_sauvignon_wine cacao  \
0        No     No         No      No                      No    No   
1        No     No         No      No                      No    No   
2        No     No         No      No                      No    No   
3        No     No         No      No                      No    No   
4        No     No         No      No                      No    No   

  camembert_cheese cane_molasses caraway cardamom carnation carob carrot  \
0               No            No      No       No        No    No    Yes   
1               No            No      No       No        No    No     No   
2               No            No      No       No        No    No     No   
3   

In [5]:
recipes.shape

(57691, 384)

The recipes shape is (57691, 384) this implies that there are 57691 recipies with 384 ingredents.

### checking for existence of ingredent
We know that a basic sushi recipe includes the ingredients:

- rice
- soy sauce
- wasabi
- some fish/vegetables

In [7]:
ingredients = list(recipes.columns.values)
print([match.group(0) for ingredient in ingredients for match in [(re.compile(".*(rice).*")).search(ingredient)] if match ] )
print([match.group(0) for ingredient in ingredients for match in [(re.compile(".*(wasabi).*")).search(ingredient)] if match])
print([match.group(0) for ingredient in ingredients for match in [(re.compile(".*(soy).*")).search(ingredient)] if match])

['brown_rice', 'licorice', 'rice']
['wasabi']
['soy_sauce', 'soybean', 'soybean_oil']


## Data preparatin stage for data modeling:
This Stage involves exploring the data futhere and making it right for the machine learning algorithm that selected in analytic approach stage.



In [8]:
recipes["country"].value_counts()

country
American        40150
Mexico           1754
Italian          1715
Italy            1461
Asian            1176
                ...  
Indonesia          12
Belgium            11
East-African       11
Israel              9
Bangladesh          4
Name: count, Length: 69, dtype: int64

## Observation
1. Cuisine column is labeled as country, which is inaccurate.
2. Cuisine names are not consistent as not all of them start with an upercase
3. Some Cuisines are duplicated as variation od country name such as vietnam and vietnamese
4. Some Cuisines have ery few recipes
   

Fix of these problems:

In [18]:
column_name = recipes.columns.values
column_name[0] = "cuisine"
recipes.columns = column_name

recipes

cuisine almond angelica anise anise_seed apple apple_brandy apricot  \
0      Vietnamese     No       No    No         No    No           No      No   
1      Vietnamese     No       No    No         No    No           No      No   
2      Vietnamese     No       No    No         No    No           No      No   
3      Vietnamese     No       No    No         No    No           No      No   
4      Vietnamese     No       No    No         No    No           No      No   
...           ...    ...      ...   ...        ...   ...          ...     ...   
57686       Japan     No       No    No         No    No           No      No   
57687       Japan     No       No    No         No    No           No      No   
57688       Japan     No       No    No         No    No           No      No   
57689       Japan     No       No    No         No    No           No      No   
57690       Japan     No       No    No         No    No           No      No   

      armagnac artemisia artichoke asparagus avocado bacon baked_potato balm  \
0           No        No        No        No      No    No           No   No   
1           No        No        No        No      No    No           No   No   
2           No        No        No        No      No    No           No   No   
3           No        No        No        No      No    No           No   No   
4           No        No        No        No      No    No           No   No   
...        ...       ...       ...       ...     ...   ...          ...  ...   
57686       No        No        No        No      No    No           No   No   
57687       No        No        No        No      No    No           No   No   
57688       No        No        No        No      No    No           No   No   
57689       No        No        No        No      No   Yes           No   No   
57690       No        No        No        No      No    No           No   No   

      banana barley bartlett_pear basil bay bean beech beef beef_broth  \
0         No     No            No   Yes  No   No    No   No         No   
1         No     No            No    No  No   No    No   No         No   
2         No     No            No    No  No   No    No   No         No   
3         No     No            No   Yes  No  Yes    No   No        Yes   
4         No     No            No    No  No   No    No   No         No   
...      ...    ...           ...   ...  ..  ...   ...  ...        ...   
57686     No     No            No    No  No   No    No   No         No   
57687     No     No            No    No  No   No    No   No         No   
57688     No    Yes            No    No  No   No    No   No         No   
57689     No     No            No    No  No   No    No   No         No   
57690     No    Yes            No    No  No   No    No   No         No   

      beef_liver beer beet bell_pepper bergamot berry bitter_orange  \
0             No   No   No          No       No    No            No   
1             No   No   No          No       No    No            No   
2             No   No   No          No       No    No            No   
3             No   No   No          No       No    No            No   
4             No   No   No          No       No    No            No   
...          ...  ...  ...         ...      ...   ...           ...   
57686         No   No   No          No       No    No            No   
57687         No   No   No          No       No    No            No   
57688         No   No   No          No       No    No            No   
57689         No   No   No          No       No    No            No   
57690         No   No   No          No       No    No            No   

      black_bean black_currant black_mustard_seed_oil black_pepper  \
0             No            No                     No           No   
1             No            No                     No          Yes   
2             No            No                     No           No   
3             No            No                     No 

## making all cusine names lowercase

In [19]:
recipes['cuisine'] = recipes['cuisine'].str.lower()

## Make the cuisine name consistent

In [22]:
recipes.loc[recipes["cuisine"] == "austria", 'cuisine'] = "austrian"

recipes.loc[recipes["cuisine"] == "belgium", "cuisine"] = "belgian"
recipes.loc[recipes["cuisine"] == "china", "cuisine"] = "chinese"
recipes.loc[recipes["cuisine"] == "canada", "cuisine"] = "canadian"
recipes.loc[recipes["cuisine"] == "netherlands", "cuisine"] = "dutch"
recipes.loc[recipes["cuisine"] == "france", "cuisine"] = "french"
recipes.loc[recipes["cuisine"] == "germany", "cuisine"] = "german"
recipes.loc[recipes["cuisine"] == "india", "cuisine"] = "indian"
recipes.loc[recipes["cuisine"] == "indonesia", "cuisine"] = "indonesian"
recipes.loc[recipes["cuisine"] == "iran", "cuisine"] = "iranian"
recipes.loc[recipes["cuisine"] == "italy", "cuisine"] = "italian"
recipes.loc[recipes["cuisine"] == "japan", "cuisine"] = "japanese"
recipes.loc[recipes["cuisine"] == "israel", "cuisine"] = "israeli"
recipes.loc[recipes["cuisine"] == "korea", "cuisine"] = "korean"
recipes.loc[recipes["cuisine"] == "lebanon", "cuisine"] = "lebanese"
recipes.loc[recipes["cuisine"] == "malaysia", "cuisine"] = "malaysian"
recipes.loc[recipes["cuisine"] == "mexico", "cuisine"] = "mexican"
recipes.loc[recipes["cuisine"] == "pakistan", "cuisine"] = "pakistani"
recipes.loc[recipes["cuisine"] == "philippines", "cuisine"] = "philippine"
recipes.loc[recipes["cuisine"] == "scandinavia", "cuisine"] = "scandinavian"
recipes.loc[recipes["cuisine"] == "spain", "cuisine"] = "spanish_portuguese"
recipes.loc[recipes["cuisine"] == "portugal", "cuisine"] = "spanish_portuguese"
recipes.loc[recipes["cuisine"] == "switzerland", "cuisine"] = "swiss"
recipes.loc[recipes["cuisine"] == "thailand", "cuisine"] = "thai"
recipes.loc[recipes["cuisine"] == "turkey", "cuisine"] = "turkish"
recipes.loc[recipes["cuisine"] == "vietnam", "cuisine"] = "vietnamese"
recipes.loc[recipes["cuisine"] == "uk-and-ireland", "cuisine"] = "uk-and-irish"
recipes.loc[recipes["cuisine"] == "irish", "cuisine"] = "uk-and-irish"

recipes

cuisine almond angelica anise anise_seed apple apple_brandy apricot  \
0      vietnamese     No       No    No         No    No           No      No   
1      vietnamese     No       No    No         No    No           No      No   
2      vietnamese     No       No    No         No    No           No      No   
3      vietnamese     No       No    No         No    No           No      No   
4      vietnamese     No       No    No         No    No           No      No   
...           ...    ...      ...   ...        ...   ...          ...     ...   
57686    japanese     No       No    No         No    No           No      No   
57687    japanese     No       No    No         No    No           No      No   
57688    japanese     No       No    No         No    No           No      No   
57689    japanese     No       No    No         No    No           No      No   
57690    japanese     No       No    No         No    No           No      No   

      armagnac artemisia artichoke asparagus avocado bacon baked_potato balm  \
0           No        No        No        No      No    No           No   No   
1           No        No        No        No      No    No           No   No   
2           No        No        No        No      No    No           No   No   
3           No        No        No        No      No    No           No   No   
4           No        No        No        No      No    No           No   No   
...        ...       ...       ...       ...     ...   ...          ...  ...   
57686       No        No        No        No      No    No           No   No   
57687       No        No        No        No      No    No           No   No   
57688       No        No        No        No      No    No           No   No   
57689       No        No        No        No      No   Yes           No   No   
57690       No        No        No        No      No    No           No   No   

      banana barley bartlett_pear basil bay bean beech beef beef_broth  \
0         No     No            No   Yes  No   No    No   No         No   
1         No     No            No    No  No   No    No   No         No   
2         No     No            No    No  No   No    No   No         No   
3         No     No            No   Yes  No  Yes    No   No        Yes   
4         No     No            No    No  No   No    No   No         No   
...      ...    ...           ...   ...  ..  ...   ...  ...        ...   
57686     No     No            No    No  No   No    No   No         No   
57687     No     No            No    No  No   No    No   No         No   
57688     No    Yes            No    No  No   No    No   No         No   
57689     No     No            No    No  No   No    No   No         No   
57690     No    Yes            No    No  No   No    No   No         No   

      beef_liver beer beet bell_pepper bergamot berry bitter_orange  \
0             No   No   No          No       No    No            No   
1             No   No   No          No       No    No            No   
2             No   No   No          No       No    No            No   
3             No   No   No          No       No    No            No   
4             No   No   No          No       No    No            No   
...          ...  ...  ...         ...      ...   ...           ...   
57686         No   No   No          No       No    No            No   
57687         No   No   No          No       No    No            No   
57688         No   No   No          No       No    No            No   
57689         No   No   No          No       No    No            No   
57690         No   No   No          No       No    No            No   

      black_bean black_currant black_mustard_seed_oil black_pepper  \
0             No            No                     No           No   
1             No            No                     No          Yes   
2             No            No                     No           No   
3             No            No                     No 

## Removing cousines with < 50 recipes

In [28]:
recipes_counts = recipes["cuisine"].value_counts()
cuisines_indices = recipes_counts > 50
cuisines_indices

cuisine
american                    True
italian                     True
mexican                     True
french                      True
asian                       True
east_asian                  True
korean                      True
canadian                    True
indian                      True
western                     True
chinese                     True
spanish_portuguese          True
uk-and-irish                True
southern_soulfood           True
jewish                      True
japanese                    True
german                      True
thai                        True
mediterranean               True
scandinavian                True
middleeastern               True
central_southamerican       True
eastern-europe              True
greek                       True
english_scottish            True
caribbean                   True
cajun_creole                True
easterneuropean_russian     True
moroccan                    True
african                     True
so

In [31]:
# give all cuisine name as number array
cuisine_name_array =recipes_counts.index.values
cuisine_name_array

array(['american', 'italian', 'mexican', 'french', 'asian', 'east_asian',
       'korean', 'canadian', 'indian', 'western', 'chinese',
       'spanish_portuguese', 'uk-and-irish', 'southern_soulfood',
       'jewish', 'japanese', 'german', 'thai', 'mediterranean',
       'scandinavian', 'middleeastern', 'central_southamerican',
       'eastern-europe', 'greek', 'english_scottish', 'caribbean',
       'cajun_creole', 'easterneuropean_russian', 'moroccan', 'african',
       'southwestern', 'south-america', 'vietnamese', 'north-african',
       'philippine', 'dutch', 'lebanese', 'iranian', 'austrian', 'swiss',
       'pakistani', 'malaysian', 'turkish', 'south-african',
       'west-african', 'indonesian', 'belgian', 'east-african', 'israeli',
       'bangladesh'], dtype=object)

In [33]:
# conver the true/faulse series to number array
cuisines_indices_array = np.array(cuisines_indices)
cuisines_indices_array

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False])

In [ ]:
# Use the boolean mask to select only cuisine with true

cuisine_name_greater_50 = np.array(cuisine_name_array)[cuisines_indices_array]
cuisine_to_keep = list(cuisine_name_greater_50)
cuisine_to_keep

In [47]:
rows_before = recipes.shape[0]
print("Number of rows of original dataframe is {}.".format(rows_before))
recipes = recipes.loc[recipes['cuisine'].isin(cuisine_to_keep)]
rows_after = recipes.shape[0]

print("Number of rows of processed dataframe is {}.".format(rows_after))

print("{} rows removed!".format(rows_before - rows_after))

Number of rows of original dataframe is 57394.
Number of rows of processed dataframe is 57394.
0 rows removed!


## Convert all Yes's to 1's and the No's to 0's

In [57]:
recipes = recipes.replace(to_replace="Yes", value = 1)
recipes = recipes.replace(to_replace="No", value = 0)
recipes["rice"]

0        1
1        0
2        0
3        1
4        1
        ..
57686    0
57687    0
57688    1
57689    0
57690    1
Name: rice, Length: 57394, dtype: int64

##### analyze the data a little more in order to learn the data better and note any interesting preliminary observations.

getting the recipes that contains rice and soy, and wasabi and seaweed

In [64]:
check_recipes = recipes.loc[(recipes["rice"] == 1) & (recipes["soy_sauce"] == 1) &
    (recipes["wasabi"] == 1) &
    (recipes["seaweed"] == 1)]
check_recipes

cuisine  almond  angelica  anise  anise_seed  apple  apple_brandy  \
11306    japanese       0         0      0           0      0             0   
11321    japanese       0         0      0           0      0             0   
11361    japanese       0         0      0           0      0             0   
12171       asian       0         0      0           0      0             0   
12385       asian       0         0      0           0      0             0   
13010       asian       0         0      0           0      0             0   
13159       asian       0         0      0           0      0             0   
13513    japanese       0         0      0           0      0             0   
13586    japanese       0         0      0           0      0             0   
13625  east_asian       0         0      0           0      0             0   
14495  east_asian       0         0      0           0      0             0   

       apricot  armagnac  artemisia  artichoke  asparagus  avocado  bacon  \
11306        0         0          0          0          0        1      0   
11321        0         0          0          0          0        0      0   
11361        0         0          0          0          0        1      0   
12171        0         0          0          0          0        0      0   
12385        0         0          0          0          0        1      0   
13010        0         0          0          0          0        1      0   
13159        0         0          0          0          0        1      0   
13513        0         0          0          0          0        0      0   
13586        0         0          0          0          0        0      0   
13625        0         0          0          0          0        0      0   
14495        0         0          0          0          0        0      0   

       baked_potato  balm  banana  barley  bartlett_pear  basil  bay  bean  \
11306             0     0       0       0              0      0    0     0   
11321             0     0       0       0              0      0    0     0   
11361             0     0       0       0              0      0    0     0   
12171             0     0       0       0              0      0    0     0   
12385             0     0       0       0              0      0    0     0   
13010             0     0       0       0              0      0    0     0   
13159             0     0       0       0              0      0    0     0   
13513             0     0       0       0              0      0    0     0   
13586             0     0       0       0              0      0    0     0   
13625             0     0       0       0              0      0    0     0   
14495             0     0       0       0              0      0    0     0   

       beech  beef  beef_broth  beef_liver  beer  beet  bell_pepper  bergamot  \
11306      0     0           0           0     0     0            0         0   
11321      0     0           0           0     0     0            0         0   
11361      0     0           0           0     0     0            0         0   
12171      0     0           0           0     0     0            0         0   
12385      0     0           0           0     0     0            0         0   
13010      0     0           0           0     0     0            0         0   
13159      0     0           0           0     0     0            0         0   
13513      0     0           0           0     0     0            0         0   
13586      0     0           0           0     0     0            0         0   
13625      0     0           0           0     0     0            0         0   
14495      0     0           0           0     0     0            0         0   

       berry  bitter_orange  black_bean  black_currant  \
11306      0              0           0              0   
11321      0              0           0              0   
11361      0              0           0             

##### count the ingredients across all recipes

In [75]:
# Sum each column

ing = recipes.iloc[:, 1:].sum(axis=0)
ing

almond        2306
angelica         1
anise          223
anise_seed      87
apple         2420
              ... 
wood            33
yam             85
yeast         3385
yogurt        1033
zucchini      1102
Length: 383, dtype: int64

In [78]:
# define each column as a pandas series
ingredient = pd.Series(ing.index.values)
ingredient

0          almond
1        angelica
2           anise
3      anise_seed
4           apple
          ...    
378          wood
379           yam
380         yeast
381        yogurt
382      zucchini
Length: 383, dtype: object

In [80]:
ingredient = pd.Series(ing.index.values, index=np.arange(len(ing)))
ingredient

0          almond
1        angelica
2           anise
3      anise_seed
4           apple
          ...    
378          wood
379           yam
380         yeast
381        yogurt
382      zucchini
Length: 383, dtype: object

In [82]:
count = pd.Series(list(ing))
count

0      2306
1         1
2       223
3        87
4      2420
       ... 
378      33
379      85
380    3385
381    1033
382    1102
Length: 383, dtype: int64

In [83]:
count = pd.Series(list(ing), index=np.arange(len(ing)))
count

0      2306
1         1
2       223
3        87
4      2420
       ... 
378      33
379      85
380    3385
381    1033
382    1102
Length: 383, dtype: int64

In [86]:
#create a dataframe with ingredient and count as columns

ing_df = pd.DataFrame(dict(ingredient=ingredient, count=count))
ing_df

ingredient  count
0        almond   2306
1      angelica      1
2         anise    223
3    anise_seed     87
4         apple   2420
..          ...    ...
378        wood     33
379         yam     85
380       yeast   3385
381      yogurt   1033
382    zucchini   1102

[383 rows x 2 columns]

In [88]:
ing_df = pd.DataFrame(dict(ingredient=ingredient, count=count))
ing_df

ingredient  count
0        almond   2306
1      angelica      1
2         anise    223
3    anise_seed     87
4         apple   2420
..          ...    ...
378        wood     33
379         yam     85
380       yeast   3385
381      yogurt   1033
382    zucchini   1102

[383 rows x 2 columns]

In [89]:
print(ing_df.to_string())

                  ingredient  count
0                     almond   2306
1                   angelica      1
2                      anise    223
3                 anise_seed     87
4                      apple   2420
5               apple_brandy     37
6                    apricot    620
7                   armagnac     11
8                  artemisia     13
9                  artichoke    391
10                 asparagus    460
11                   avocado    660
12                     bacon   2169
13              baked_potato      9
14                      balm      3
15                    banana    989
16                    barley    266
17             bartlett_pear     23
18                     basil   3842
19                       bay   1463
20                      bean   1992
21                     beech      1
22                      beef   4902
23                beef_broth    845
24                beef_liver     10
25                      beer    307
26                      beet

#### Now sorting the dataframe in decending order of count

In [91]:
ing_df.sort_values(by='count', ascending=False, inplace=True)
ing_df.reset_index(drop=True, inplace=True)
ing_df

ingredient  count
0                 egg  21022
1               wheat  20775
2              butter  20715
3               onion  18078
4              garlic  17351
..                ...    ...
378  roasted_hazelnut      1
379   sturgeon_caviar      1
380    strawberry_jam      1
381          angelica      1
382            durian      0

[383 rows x 2 columns]

###  The data is bias towards America ingridetnt because there are about 40000 America recipe

Creating profire for each cuisine

In [98]:
cuisines = recipes.groupby("cuisine").mean()
cuisines.head()

almond  angelica     anise  anise_seed     apple  \
cuisine                                                            
african       0.156522  0.000000  0.000000    0.000000  0.034783   
american      0.040598  0.000025  0.003014    0.000573  0.052055   
asian         0.007544  0.000000  0.000838    0.002515  0.012573   
cajun_creole  0.000000  0.000000  0.000000    0.000000  0.006849   
canadian      0.036176  0.000000  0.000000    0.000000  0.036176   

              apple_brandy   apricot  armagnac  artemisia  artichoke  \
cuisine                                                                
african           0.000000  0.069565    0.0000        0.0   0.000000   
american          0.000623  0.011308    0.0001        0.0   0.006351   
asian             0.000000  0.005029    0.0000        0.0   0.000000   
cajun_creole      0.000000  0.000000    0.0000        0.0   0.000000   
canadian          0.000000  0.002584    0.0000        0.0   0.001292   

              asparagus   avocado     bacon  baked_potato      balm    banana  \
cuisine                                                                         
african        0.000000  0.008696  0.008696      0.000000  0.000000  0.017391   
american       0.007522  0.007323  0.041220      0.000199  0.000075  0.021594   
asian          0.019279  0.010897  0.013412      0.000000  0.000000  0.006706   
cajun_creole   0.000000  0.006849  0.020548      0.000000  0.000000  0.006849   
canadian       0.007752  0.009044  0.032300      0.000000  0.000000  0.020672   

                barley  bartlett_pear     basil       bay      bean  beech  \
cuisine                                                                      
african       0.017391       0.000000  0.017391  0.034783  0.008696    0.0   
american      0.003885       0.000423  0.058655  0.020697  0.028767    0.0   
asian         0.029338       0.000000  0.055323  0.011735  0.083822    0.0   
cajun_creole  0.000000       0.000000  0.027397  0.232877  0.006849    0.0   
canadian      0.007752       0.000000  0.060724  0.028424  0.019380    0.0   

                  beef  beef_broth  beef_liver      beer      beet  \
cuisine                                                              
african       0.052174    0.008696    0.000000  0.000000  0.008696   
american      0.075118    0.009988    0.000174  0.005430  0.003686   
asian         0.103940    0.012573    0.000000  0.004191  0.000838   
cajun_creole  0.143836    0.027397    0.000000  0.000000  0.013699   
canadian      0.086563    0.009044    0.000000  0.010336  0.002584   

              bell_pepper  bergamot     berry  bitter_orange  black_bean  \
cuisine                                                                    
african          0.313043  0.000000  0.000000       0.000000    0.008696   
american         0.087846  0.000149  0.003836       0.001245    0.005131   
asian            0.133277  0.000000  0.000838       0.001676    0.017603   
cajun_creole     0.342466  0.000000  0.000000       0.000000    0.000000   
canadian         0.100775  0.000000  0.003876       0.001292    0.007752   

              black_currant  black_mustard_seed_oil  black_pepper  \
cuisine                                                             
african            0.000000                0.000000      0.295652   
american           0.000199                0.000050      0.147621   
asian              0.000000                0.001676      0.118189   
cajun_creole       0.000000                0.000000      0.328767   
canadian           0.000000                0.001292      0.209302   

              black_raspberry  black_sesame_seed  black_tea  blackberry  \
cuisine                                                                   
african              0.000000                0.0   0.000000    0.000000   
american             0.000199                0.0   0.000573    0.003462   
asian                0.000000                0.0   0.005868    0.000838   
cajun_creole         0.000000       

In [105]:
num_ingredient = 4   # define number of top ingredient to print

# define a function that prints the top ingredient for each cuisine

def print_top_ingredients(row):
    print(row.name.upper())
    row_sorted = row.sort_values(ascending=False)*100
    top_ingredients = list(row_sorted.index.values[0:num_ingredient])

    for ind, ingridient in enumerate(top_ingredients):
        print("%s(%d%%) " % (ingridient, row_sorted[ind]), end='')
    print("\n" )

create_cuisines_profile = cuisines.apply(print_top_ingredients, axis=1)


AFRICAN
onion(53%) olive_oil(52%) garlic(49%) cumin(42%) 

AMERICAN
butter(41%) egg(40%) wheat(39%) onion(29%) 

ASIAN
soy_sauce(49%) ginger(48%) garlic(47%) rice(41%) 

CAJUN_CREOLE
onion(69%) cayenne(56%) garlic(48%) butter(36%) 

CANADIAN
wheat(39%) butter(38%) egg(35%) onion(34%) 

CARIBBEAN
onion(51%) garlic(50%) vegetable_oil(31%) black_pepper(31%) 

CENTRAL_SOUTHAMERICAN
garlic(56%) onion(54%) cayenne(51%) tomato(41%) 

CHINESE
soy_sauce(68%) ginger(53%) garlic(52%) scallion(48%) 

EAST_ASIAN
garlic(55%) soy_sauce(50%) scallion(49%) cayenne(47%) 

EASTERN-EUROPE
wheat(53%) egg(52%) butter(48%) onion(45%) 

EASTERNEUROPEAN_RUSSIAN
butter(60%) egg(50%) wheat(49%) onion(38%) 

ENGLISH_SCOTTISH
butter(67%) wheat(62%) egg(53%) cream(41%) 

FRENCH
butter(50%) egg(44%) wheat(37%) olive_oil(27%) 

GERMAN
wheat(64%) egg(60%) butter(47%) onion(34%) 

GREEK
olive_oil(76%) garlic(44%) onion(36%) lemon_juice(33%) 

INDIAN
cumin(60%) turmeric(50%) onion(49%) coriander(47%) 

ITALIAN
olive_oil

C:\Users\HP\AppData\Local\Temp\ipykernel_19256\1534779729.py:11: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print("%s(%d%%) " % (ingridient, row_sorted[ind]), end='')


At this point, we feel that we have understood the data well and the data is ready and is in the right format for modeling!